### PR TITLE
sql: prevent dropping a sequence if it's being used

### PR DESCRIPTION
**Pull request by vilterp**
Save and update references between sequence descriptor and column descriptor; check these references when trying to drop a sequence.

Addresses #20964

### CODE REVIEW

**Comment by vilterp** [comment]
Requesting review from @jordanlewis and @vivekmenezes because it does a lot of modification of table descriptors, and I may have made mistakes with how those changes are saved / notified.

**Comment by jordanlewis** [comment]
Taking a look at this today. Sorry for the delay!

**Comment by jordanlewis** [comment]
Generally looks good! The main question on my mind is about defaults that have more than one `nextval` call which doesn't seem to be properly handled.

---

Reviewed 5 of 8 files at r1, 1 of 2 files at r2, 1 of 3 files at r3, 2 of 2 files at r5, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sequence.go, line 241 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35eSuJJmi72oKso9Fb:-L35eSuKONPPejBIO0Pw:b-n8h7tp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/sequence.go#L241)):*
> ```Go
> 	}
> 	col.UsesSequenceId = &seqDesc.ID
> 	// Now that we've gotten the sequence descriptor, I guess we can mutate it?
> ```

Probably should change this comment to be less questiony :)

---

*[pkg/sql/sequence.go, line 289 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35eYY7_m7V_P0W2Z28:-L35eYY7_m7V_P0W2Z29:bo3j0vf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/sequence.go#L289)):*
> ```Go
> // no call to nextval.
> // e.g. nextval('foo') => "foo"; <some other expression> => nil
> func getUsedSequenceName(defaultExpr tree.TypedExpr) (*string, error) {
> ```

What if there is more than one nextval in the expression?

---

*[pkg/sql/logictest/testdata/logic_test/sequences, line 597 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35enpgbP3uSTaLyijL:-L35enpgbP3uSTaLyijM:b-ygdpt6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/logictest/testdata/logic_test/sequences#L597)):*
> ```
> 
> statement ok
> ALTER TABLE drop_default_tesT_tbl ALTER COLUMN id SET DEFAULT nextval('drop_default_test_seq_2')
> ```

What about some tests that we can drop the sequence after the entire table is dropped, not just the column?

---

*[pkg/sql/sqlbase/structured.proto, line 131 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35ebUug8PkIy4eZN-9:-L35ebUug8PkIy4eZN-A:bpbv2ko) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/sqlbase/structured.proto#L131)):*
> ```protobuf
>   // Id of a sequence used in this column's DEFAULT expression, in a call to nextval().
>   // null if no sequence is used.
>   optional uint32 uses_sequence_id = 10 [(gogoproto.nullable) = true, (gogoproto.casttype) = "ID"];
> ```

Probably needs to be a repeated field.

---

*[pkg/sql/sqlbase/table.go, line 157 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35eeZCangp2A97vsUi:-L35eeZCangp2A97vsUj:b-xokrx4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/sqlbase/table.go#L157)):*
> ```Go
> func MakeColumnDefDescs(
> 	d *tree.ColumnTableDef, semaCtx *tree.SemaContext, evalCtx *tree.EvalContext,
> ) (*ColumnDescriptor, *IndexDescriptor, tree.TypedExpr, error) {
> ```

Please add a comment to explain the new return parameter.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-:-L35f-Zr8wrjj43AlDTd:bn6vji1)*

File: pkg/sql/sequence.go [hunk 1]
```diff
@@ -208,3 +209,109 @@ func assignSequenceOptions(
...
+// - if the given expr uses a sequence
+//   - adds a reference on `col` to the sequence. (mutates `col`; caller must persist changes to KV)
+//   - adds a reference on the sequence descriptor to `col`. (returns the modified sequence
+//     descriptor for the caller to save)
+func maybeAddSequenceDependency(
+	tableDesc *sqlbase.TableDescriptor,
+	col *sqlbase.ColumnDescriptor,
+	expr tree.TypedExpr,
+	evalCtx *tree.EvalContext,
+) (*sqlbase.TableDescriptor, error) {
```
Comment (vivekmenezes) [inline pkg/sql/sequence.go]: Can you add a comment on what descriptor is returned?

File: pkg/sql/sequence.go [hunk 2]
```diff
@@ -208,3 +209,109 @@ func assignSequenceOptions(
...
+	}
+	seqDesc, err := getSequenceDesc(ctx, evalCtx.Txn, NilVirtualTabler, parsedSeqName)
+	if err != nil {
+		return nil, err
+	}
+	col.UsesSequenceId = &seqDesc.ID
+	// Now that we've gotten the sequence descriptor, I guess we can mutate it?
+	seqDesc.DependedOnBy = append(seqDesc.DependedOnBy, sqlbase.TableDescriptor_Reference{
+		ID:        tableDesc.ID,
+		ColumnIDs: []sqlbase.ColumnID{col.ID},
```
Comment (vivekmenezes) [inline pkg/sql/sequence.go]: what if there are multiple columns using the same sequence?

File: pkg/sql/sequence.go [hunk 3]
```diff
@@ -208,3 +209,109 @@ func assignSequenceOptions(
...
+	tableDesc *sqlbase.TableDescriptor, col *sqlbase.ColumnDescriptor, params runParams,
+) error {
+	// Get the sequence descriptor so we can remove the reference from it.
+	seqDesc := sqlbase.TableDescriptor{}
+	if err := getDescriptorByID(params.ctx, params.p.txn, *col.UsesSequenceId, &seqDesc); err != nil {
+		return err
+	}
+	// Find an item in seqDesc.DependedOnBy which references tableDesc.
+	refIdx := -1
+	for i := 0; i < len(seqDesc.DependedOnBy); i++ {
```
Comment (vivekmenezes) [inline pkg/sql/sequence.go]: for i, reference := range ...

File: pkg/sql/sqlbase/structured.proto [hunk 4]
```diff
@@ -126,6 +126,9 @@ message ColumnDescriptor {
   reserved 9;
   optional bool hidden = 6 [(gogoproto.nullable) = false];
   reserved 7;
+  // Id of a sequence used in this column's DEFAULT expression, in a call to nextval().
+  // null if no sequence is used.
+  optional uint32 uses_sequence_id = 10 [(gogoproto.nullable) = true, (gogoproto.casttype) = "ID"];
```
Comment (vivekmenezes) [inline pkg/sql/sqlbase/structured.proto]: Is it possible for more than one sequence to be used?

**Review by vivekmenezes — COMMENTED** [review]
LGTM

File: pkg/sql/sequence.go (hunk 3 above)
Comment (vilterp) [reply pkg/sql/sequence.go]: Done

File: pkg/sql/sequence.go (hunk 2 above)
Comment (vilterp) [reply pkg/sql/sequence.go]: `maybeAddSequenceDependency` gets called for each column, so it'll end up as a separate entry in `seqDesc.DependedOnBy` for each. Will add a logic test for this.

**Comment by vilterp** [comment]
TFTRs. PTAL

---

Review status: 1 of 15 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sequence.go, line 241 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35eSuJJmi72oKso9Fb:-L3A4pHEgCbk6lVVYJ9j:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/sequence.go#L241)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Probably should change this comment to be less questiony :)
</blockquote></details>

Done.

---

*[pkg/sql/sequence.go, line 289 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35eYY7_m7V_P0W2Z28:-L3A4q_r0xisa6R5OSi6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/sequence.go#L289)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

What if there is more than one nextval in the expression?
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/sequences, line 597 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35enpgbP3uSTaLyijL:-L3A4rZ3vdKqhXgjZtvS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/logictest/testdata/logic_test/sequences#L597)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

What about some tests that we can drop the sequence after the entire table is dropped, not just the column?
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/structured.proto, line 131 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35ebUug8PkIy4eZN-9:-L3A4sOrRIulRsgeOK0K:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/sqlbase/structured.proto#L131)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Probably needs to be a repeated field.
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/table.go, line 157 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35eeZCangp2A97vsUi:-L3A4tYQ_CzoD_DqFmtw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/sqlbase/table.go#L157)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Please add a comment to explain the new return parameter.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-:-L3A4vqeWEZwuyQyuhU9:bqslf27)*

**Comment by knz** [comment]
nit: it's frankly incorrect to not have entered any release note in any of these commits. Here a pretty significant bug was fixed and it would have deserved mention in the next batch of release notes.

Adding the label `docs-todo` as a reminder to have this PR properly documented in the next release notes despite the lack of release note annotation. cc @jseldess .

**Comment by jseldess** [comment]
Thanks, @knz. Tracked here: https://github.com/cockroachdb/docs/issues/2409

**Pull request merged**